Add path specification to scaffolds

jtoar opened the pull request:
This PR proposes the following change to the `g scaffold` command:

```
yarn rw g scaffold <path/model>
```

E.g. given the model user, running `yarn rw g scaffold admin/user` will output:

```
yarn run v1.22.4
$ /home/dominic/projects/redwood/redwood-test/node_modules/.bin/rw g scaffold admin/users
  ✔ Generating scaffold files...
    ✔ Writing `./api/src/graphql/users.sdl.js`...
    ✔ Writing `./api/src/services/users/users.test.js`...
    ✔ Writing `./api/src/services/users/users.js`...
    ✔ Writing `./web/src/scaffold.css`...
    ✔ Writing `./web/src/layouts/UsersLayout/UsersLayout.js`...
    ✔ Writing `./web/src/pages/Admin/EditUserPage/EditUserPage.js`...
    ✔ Writing `./web/src/pages/Admin/UserPage/UserPage.js`...
    ✔ Writing `./web/src/pages/Admin/UsersPage/UsersPage.js`...
    ✔ Writing `./web/src/pages/Admin/NewUserPage/NewUserPage.js`...
    ✔ Writing `./web/src/components/Admin/EditUserCell/EditUserCell.js`...
    ✔ Writing `./web/src/components/Admin/User/User.js`...
    ✔ Writing `./web/src/components/Admin/UserCell/UserCell.js`...
    ✔ Writing `./web/src/components/Admin/UserForm/UserForm.js`...
    ✔ Writing `./web/src/components/Admin/Users/Users.js`...
    ✔ Writing `./web/src/components/Admin/UsersCell/UsersCell.js`...
    ✔ Writing `./web/src/components/Admin/NewUser/NewUser.js`...
  ✔ Adding scaffold routes...
  ✔ Adding scaffold asset imports...
Done in 1.07s.
```

## Motivation

Scaffolds are amazing and could be a bit more amazing. The knock against scaffolds this PR addresses is their tendency to pollute the namespace.

The specific problem I had was suddenly having to contrive names for my app's real components and pages. I.e. I had a model called `notes`, and scaffolding this out gave me a CRUD, but key names were taken: `Note` and `Notes` as components, `NotePage` and `NotesPages` as a pages, etc. I wanted to keep the scaffold as a CRUD while fleshing things out, but having to manually change everything seemed like a lot of work that I thought redwood should've made trivial from the get-go. Specifying a path can make the kind of admin organization seen in [example-blog](https://github.com/redwoodjs/example-blog) possible right away. 

There's an understandably tight link between a model's name and the scaffold generated. But this coupling must be loosened a bit. Specifying a path is one way of doing this. Another that's been thrown around involves a `--prefix` flag to literally add a prefix to component and page names. 

> This PR stems from discussions on discourse: see [How to Think About Scaffolds](https://community.redwoodjs.com/t/how-to-think-about-scaffolds/254) and [The scaffolding command makes it hard to delete files and folders later](https://community.redwoodjs.com/t/the-scaffolding-command-makes-it-hard-to-delete-files-and-folders-later/320).

I'll continue to scheme/push changes, but wanted to get this out to start a discussion/get help before moving too far in any one direction. Thanks/please help @cannikin @thedavidprice @olance.

Questions
* The way I finagled yargs command seems very suspect?
* Maybe I should just make another Listr task to move it all after it's been generated? Right now I'm making changes to the first task (`Generating scaffold files...`).
* Should the generated layout be in a nested dir too? Can it even be?

olance commented:
That looks quite nice to me!
You should probably add some tests to ensure things are created with the correct paths, names and imports for a few different cases.

As for the command-line itself, I'm wondering whether going for `path/model` is the "right thing" to do. It certainly works, but in terms of semantics and use case I wonder if a `--prefix` or `--path` option wouldn't make more sense?

My reasoning is the following (hopefully I make it clear enough):
- As a developer, I'm using this generator to create a new model for my app, along with useful ready-to-use CRUD pages. Which means it's the _model_ that I'm mainly interested in. As you mentioned yourself, I might want to get rid of those scaffolds later, or hide them behind an admin login... but the model itself would remain.
- Said differently, I am not interested in creating a "admin user" model, but a "user" model for which I want CRUD pages to be considered admin pages

I think this kinda shows in the fact that everything generated on the `api` side doesn't use the `admin` prefix, whereas everything on the `web` side does.

So with that logic in mind, I'm thinking `yarn rw g scaffold --path admin user` might be a better candidate. wdyt?

I'd still use a namespacing via nested paths as you did though, rather than prefixing the classes names. Otherwise I feel the `components` and `pages` directories would get cluttered very quickly.

jtoar commented:
@olance quick clarification:

> As a developer, I'm using this generator to create a new model for my app, along with useful ready-to-use CRUD pages

I think `yarn rw g scaffold` doesn't actually create a new model. I.e. if I run `yarn rw g scaffold user` without having a `model User` in my `schema.prisma` the command will fail:

```
Cannot create property 'context' on string 'No schema definition found for `User` in schema.prisma file'
```

> Said differently, I am not interested in creating a "admin user" model, but a "user" model for which I want CRUD pages to be considered admin pages

That's exactly what I'm interested too. I'll quote @cannikin here (pulling from the discourse thread)--I was basically implementing his suggestion since I thought it more declarative:

> One way may be to simply include a path where you want the generated files to live right in the name of the model itself:
>
> yarn rw g scaffold admin/post
>
> The last part of the path would have to match a model in your database, but everything before that would be up to you. This would put pages, components and routes in the admin subdirectory. So you still end up with PostsCell but when importing it it would be in src/components/admin/PostsCell.

A flag was what I originally had in mind, but it was a `--prefix` flag instead of a `--path` flag. I'm ready to put the `--prefix` flag to rest as, as you mentioned at the end, nested paths > prefix.

olance commented:
> I think `yarn rw g scaffold` doesn't actually create a new model. I.e. if I run `yarn rw g scaffold user` without having a `model User` in my `schema.prisma`, the command will fail:

Oooh my bad... Rails scaffold generator does create the model if it's missing, so I assumed Redwood's one would do the same!

@cannikin what's your take on the `--path` versus `admin/post` approach?

thedavidprice commented:
FYI: we’ve added a CI runner for Windows and changed PR settings to require checks to pass before merging. Pulling in changes from master should resolve current warnings.

jtoar commented:
@thedavidprice Thank you, resolved!

thedavidprice commented:
@cannikin this one’s ready for review

thedavidprice commented:
@jtoar just checking back on this to make sure it doesn’t get lost in the shuffle. No urgency implied.

But let us know if you’d like to take on the tests and need any help doing so.

jtoar commented:
@thedavidprice thanks, I've started taking on the tests but could use a little feedback @cannikin. Here's a quick update:

- Path nesting can be > 1 (i.e. `super/admin/user` will work)
- I added some tests
  - Do I need to make all these `fixtures` dirs? Is there a better way?
- I had to make a change in redwood's `internal` package
---

I added some tests (`scaffoldPathSlashModel.test.js`, `scaffoldPathSlashModelNested.test.js`) for the following inputs:
 - `admin/post`
 - `super/admin/post`
   
These tests are me literally copying your scaffold.test.js and adapting it to the new input. Which means I made more `fixtures` directories (`fixturesPathSlashModel`, `fixturesPathSlashModelNested`). Is this the best way to do it? Because if I test the other inputs you mentioned (`admin_pages/post`, `Admin/Post`, `AdminPages/Post`), I'll be making three more `fixtures` that, like the previous two, aren't really that different from each other (the difference mainly being in the import statements).

---

I couldn't get doubly-nested paths like `Super/Admin/Post` to be automatically imported correctly into in Routes.js.

<details><summary>The error I was getting:</summary><p>

![routes-error](https://user-images.githubusercontent.com/32992335/80695572-f5ab8500-8a8a-11ea-9613-967fcc3c0fd5.png)

</p></details>

I applied a fix in the `internal` package's `paths.ts`. It's definitely more of a hack and should probably be thought through a little more.

cannikin commented:
@jtoar Yeah that's getting to be a lot of fixtures. I'd be happy with just testing that the file was created at the correct location, with a test like:

```js
expect(files).toHaveProperty([
  '/path/to/project/web/src/pages/Admin/UsersPage/UsersPage.js',
])
```

And then a test that the file content contains the expected import statement. Something like:

```js
expect(files['/path/to/project/web/src/pages/Admin/UsersPage/UsersPage.js']).toMatch(`import UsersCell from 'src/components/Admin/UsersCell'`)
```

How does that feel? You can get rid of all those copied fixture files, sorry about that!

cannikin commented:
@peterp Any ideas about this?

> I couldn't get doubly-nested paths like `Super/Admin/Post` to be automatically imported correctly into in Routes.js.
> 
> The error I was getting:
> ![image](https://user-images.githubusercontent.com/300/80772131-7d84a400-8b0a-11ea-850f-6ac917a84f3b.png)
> I applied a fix in the `internal` package's `paths.ts`. It's definitely more of a hack and should probably be thought through a little more.

jtoar commented:
@cannikin Tests are here!

- `scaffoldPath.test.js` tests the `Admin/Post` case (i.e. just one level of nesting)
- `scaffoldPathMultiword.test.js` test the `AdminPages/Post` case (variations on a one-level nest)
- `scaffoldPathNested.test.js` tests the `Admin/Pages/Post` case (two levels)

The structure of the tests are largely the same, so at first I tried to save my fingers by using jest's `test.each` / for loops / _any_ kind of iteration, but I just couldn't get anything to work. Hence the +2.5k lines. :grimacing: 

I need to do more research, but based one the way our `beforeAll`'s set up, it looks like jest just doesn't support it. The last line of [jest's troubleshooting](https://jestjs.io/docs/en/troubleshooting) page:

> Note: This means when you are using test.each you cannot set the table asynchronously within a beforeEach / beforeAll.

And I think that's what we do. So even though the tests seem to work, I'd be interested in knowing if there's a better way I could've done them!

---

@peterp The fix is in file `packages/internal/src/paths.ts`. Here's the (poorly reproduced) diff for convenience:

```javascript
- importStatement: `const ${importName} = { name: '${importName}', loader: () => import('${importFile}') }`,
+ importStatement: `const ${importName.split(',').join('')} = { name: '${importName.split(',').join('')}', loader: () => import('${importFile}') }`,
```

peterp commented:
@jtoar Got it, so `importName` changed from something like this: `'Admin,Pages,Post'`,  to this `'AdminPagesPost'`

It looks good to me.

jtoar commented:
@peterp Exactly. Routes.js has a preamble saying you can nest components in however many directories you want as long as you prefix the component with the directory names:

```javascript
// In this file, all Page components from 'src/pages` are auto-imported. Nested
// directories are supported, and should be uppercase. Each subdirectory will be
// prepended onto the component name.
//
// Examples:
//
// 'src/pages/HomePage/HomePage.js'         -> HomePage
// 'src/pages/Admin/BooksPage/BooksPage.js' -> AdminBooksPage
```

I found this to be not the case when it was nested in > 1 directory. I.e. if it was 

```javascript
// 'src/pages/Super/Admin/BooksPage/BooksPage.js' -> SuperAdminBooksPage
```

It wouldn't work, the error being:

`Super,AdminBooksPage`

thedavidprice commented:
@jtoar is this ready for review? If so, please loop in Rob for the final check. Thanks!

jtoar commented:
@cannikin This one's ready for another review. The fixture-less tests have been added and @peterp okayed the change made in internals.

One last thing I wanted to make sure of: everyone prefers this syntax? The path/model syntax:

```yarn rw g scaffold admin/posts```

Instead of the path flag, which would be 

```yarn rw g scaffold posts --path=admin```

thedavidprice commented:
+1 for path/model as much more intuitive re: DX. I have to think about --path=admin

Haven't looked at the code. But would it be too much to accept either?

jtoar commented:
Great! And I don't think that'd be too hard; I'll work on that in the mean time.

jtoar commented:
@thedavidprice Added (tests still pass).

I wrote it so that if the path's specified by both `pathSlashModel` and the path flag, the path flag will prevail since it's more explicit. Is that the design we want?

thedavidprice commented:
@jtoar I think that's great! Thanks for taking the time to figure that part out. For a next step after this PR, we are going to need documentation for this.
- Yargs does allow a 'description' for options, but we need to fix something with our Yargs directory structure to allow help text to display for specific `rw generate` types. E.g. `yarn rw g scaffold --help` displays the same output as `yarn rw g --help`. 
- We need to add full Documentation for CLI commands and add to the Reference section of Website. Aryan started this in #310 and we weren't quite ready for it. But seems like it might be a good time to pick up again.

@cannikin this one looks good to me! However, I haven't run any local tests myself. And/or checked the new Jest tests.

jtoar commented:
@thedavidprice I can't get enough docs so I'll start the conversation up again over there.

@cannikin I haven't changed the routes' names, so even if you generate the scaffold with something like `admin/user`, the names still revolve around the model. E.g.:

```
<Route path="/admin/users" page={AdminUsersPage} name="users" />
```

Should the name now be `adminUsers`?

cannikin commented:
@jtoar 

> Should the name now be adminUsers?

I would expect so, yes. The `name` should get namespaced just like the directory structure and the name of the component itself.

jtoar commented:
@cannikin The route names are configured correctly now. 

I also tried to organize a bit--squashed/split everything into four commits. Sorry if I'm tripping over git while doing this (if I'm inconveniencing you guys please let me know!)

**Summary**

Just to summarize the work to date, you can now do something like:

```
yarn rw g scaffold admin/user
```
and/or
```
yarn rw g scaffold user --path=admin
```

and get this output:

```terminal
~/redwood-app$ yarn rw g scaffold admin/user
yarn run v1.22.4
$ ~/redwood-app/node_modules/.bin/rw g scaffold admin/user
  ✔ Generating scaffold files...
    ✔ Writing `./api/src/graphql/users.sdl.js`...
    ✔ Writing `./api/src/services/users/users.test.js`...
    ✔ Writing `./api/src/services/users/users.js`...
    ✔ Writing `./web/src/scaffold.css`...
    ✔ Writing `./web/src/layouts/Admin/UsersLayout/UsersLayout.js`...
    ✔ Writing `./web/src/pages/Admin/EditUserPage/EditUserPage.js`...
    ✔ Writing `./web/src/pages/Admin/UserPage/UserPage.js`...
    ✔ Writing `./web/src/pages/Admin/UsersPage/UsersPage.js`...
    ✔ Writing `./web/src/pages/Admin/NewUserPage/NewUserPage.js`...
    ✔ Writing `./web/src/components/Admin/EditUserCell/EditUserCell.js`...
    ✔ Writing `./web/src/components/Admin/User/User.js`...
    ✔ Writing `./web/src/components/Admin/UserCell/UserCell.js`...
    ✔ Writing `./web/src/components/Admin/UserForm/UserForm.js`...
    ✔ Writing `./web/src/components/Admin/Users/Users.js`...
    ✔ Writing `./web/src/components/Admin/UsersCell/UsersCell.js`...
    ✔ Writing `./web/src/components/Admin/NewUser/NewUser.js`...
  ✔ Adding scaffold routes...
  ✔ Adding scaffold asset imports...
Done in 1.14s.
```

Basically the `sdl`, `services`, and `scaffold.css` are left unaffected while `layouts`, `pages`, and `components`, are nested in dirs.

<details><summary><em>Expand to see sample tree output of web/src/</em></summary>

```terminal
web/src/
├── components
│   └── Admin
│       ├── EditUserCell
│       │   └── EditUserCell.js
│       ├── NewUser
│       │   └── NewUser.js
│       ├── User
│       │   └── User.js
│       ├── UserCell
│       │   └── UserCell.js
│       ├── UserForm
│       │   └── UserForm.js
│       ├── Users
│       │   └── Users.js
│       └── UsersCell
│           └── UsersCell.js
├── index.css
├── index.html
├── index.js
├── layouts
│   └── Admin
│       └── UsersLayout
│           └── UsersLayout.js
├── pages
│   ├── Admin
│   │   ├── EditUserPage
│   │   │   └── EditUserPage.js
│   │   ├── NewUserPage
│   │   │   └── NewUserPage.js
│   │   ├── UserPage
│   │   │   └── UserPage.js
│   │   └── UsersPage
│   │       └── UsersPage.js
│   ├── FatalErrorPage
│   │   └── FatalErrorPage.js
│   └── NotFoundPage
│       └── NotFoundPage.js
├── Routes.js
└── scaffold.css
```
</details>

I'm not sure if it'll add much value, but here's an [example repo](https://github.com/jtoar/redwood-scaffold-pr) that has the output of running this modified scaffold command. 

Quick summary of the changes I made to `scaffold.js`

1) Modifying the files/routes functions to accept a new
(optional) parameter: `scaffoldPath`

2) Defining a lot of constants like "pascalScaffoldPath" or
"camelScaffoldPath" to plug in to the templates/routes.

**Tests**

I've added three tests that check to see if files are 1) named correctly, 2) imported correctly, and 3) "routed" correctly. They mostly reuse tests from the existing `scaffold.test.js` with some modifications (i.e. they're fixture-less). The tests cover three scenarios:

- One-level nesting (`scaffoldPath.test.js`): `admin/post`
- Multiword one-level nesting (`scaffoldPathMultiword.test.js`): `adminPages/post`
- Two-level nesting (`scaffoldPathNested.test.js`): `admin/pages/post`

You can run the tests by navigating to the `@redwoodjs/cli` package and running

```
yarn test scaffold
```

I think we're getting close to the end? I'll keep looking the code over for mistakes/better ways to do htings, but let me know if I can do anything else and thanks for your feedback!

thedavidprice commented on packages/cli/src/commands/generate/scaffold/scaffold.js:
```diff
@@ -221,25 +313,42 @@ const addScaffoldImport = () => {
-export const handler = async ({ model, force }) => {
+// The user can also specify a path in the argument.
+// E.g. yarn rw g scaffold admin/post
+export const handler = async ({ pathSlashModel, force, path: pathFlag }) => {
```
possible to add a Yargs 'description' key in here? It won't display correctly yet, e.g. `yarn rw g scaffold --help` currently shows help for `yarn rw g` only. But it helps code now and will be available via CLI in future once we re-structure Yargs:
https://yargs.js.org/docs/#api-optionskey-opt

jtoar replied:
For the path flag, how does `path to put generated files in` sound? Alternatives:
- `path to generated files`
- `path to nest generated files under`
- `directory(ies) to nest generated files under`
  
And the command's description could change from/to `Generates pages, SDL, and a services object` => `Generates and optionally nests pages, SDL, and a services object`

thedavidprice commented:
Just ran through this locally -- it's really great behavior! I used the `admin/user` option, which is intuitive. And then also created `admin/contact`. Everything structured accordingly. And I tested `admin/foo/user` as well... works great.

This really needs docs, especially because the `yarn rw g scaffold --help` is broken. I added one inline review note about Yargs description key:value for Options. But for CLI documentation, this would be a good prompt to start chipping away a piece at a time. (Note: don't let docs block this from merging. But lets at least start a new Issue if we haven't already.)

@cannikin I did not review the code, just the behavior. Handing off to you!

jtoar commented:
@thedavidprice reached out about docs at #310; will start a google doc. Stay tuned for an issue most likely later today.

cannikin commented:
Great work! This has gotta be the biggest PR we've merged, 2,000+ lines 😮

jmreidy commented:
Love this! Can't wait to put it to use. Thanks @jtoar !
